library/x509_crt.c

```diff
@@ -2951,23 +2951,159 @@ static int x509_crt_verify_chain(
     }
 }
 
+/*
+ * parse ipv4 address from canonical string form into bytes.
+ * return 1 if success, 0 otherwise (consistent with inet_pton)
+ */
+static int parse_ipv4( const char *h, char *addr )
+{
+    unsigned char ip[4];
+    int i;
+    const char *strt = h;
+    char *endp;
+    unsigned long v;
+    for( i = 0; i < 4; i++ )
+    {
+        v = strtoul( strt, &endp, 10 );
+        if( endp == strt || v > 255 )
+        {
+            return ( 0 );
+        }
+
+        if( i < 3 && *endp != '.' )
+        {
+            return ( 0 );
+        }
+        ip[i] = v & (unsigned char)0xFF;
+        strt = endp + 1;
+    }
+
+    memcpy( addr, ip, sizeof( ip ) );
+    return ( 1 );
+}
+
+/*
+ * parse ipv6 address from canonical string form into bytes.
```

Comment:
I think it would be good to be more precise about what "canonical string" means here. It looks like the intention is to accept any of the first two formats described in [RFC 4291 section 2.2](https://tools.ietf.org/html/rfc4291#section-2.2), and not enforce that only the format recommended in RFC 5952 is used. I think that's a good choice but it should be make explicit.